### ISSUE TITLE
Document the intuition behind the optimized ruleset

### ISSUE THREAD

**Issue by ecstatic-morse** [rust-lang/polonius#175]
The "optimized" variant of the borrow-checking rules—[defined here](https://github.com/rust-lang/polonius/blob/master/polonius-engine/src/output/datafrog_opt.rs)—isn't explained in detail anywhere in this repo AFAICT. You can read through the relations one-by-one to get a sense of what's happening, but it's easy to miss the forest for the trees so-to-speak. For example, what is so special about "dying regions"? The book only has a short placeholder section:

> A more detailed description of the rules in this Opt variant will be added later but it computes the same data as the Naive variant described above, more efficiently, by limiting where the subset transitive closure is computed: some origins are short-lived, or part of a subsection of the subset graph into which no loan ever flows, and therefore don't contribute to errors or loan propagation. There's no need to track these specific cases.

Niko is pretty busy, so I don't foresee him writing another sweet blog post [like this one](https://smallcultfollowing.com/babysteps/blog/2018/04/27/an-alias-based-formulation-of-the-borrow-checker/#an-alias-based-formulation-of-the-borrow-checker). There's actually a pretty good explanation in the first paragraph of [one of Niko's comments](https://github.com/rust-lang/polonius/issues/153#issuecomment-669423499) on #153, but obviously that needs to be expanded on.

I've developed my own intuition about what the optimized variant is doing, and @lqd has some insight as well. It should be enough to improve the state of things. During the sprint, [Niko mentioned removing the optimized variant](https://rust-lang.zulipchat.com/#narrow/stream/186049-t-compiler.2Fwg-polonius/topic/Polonius.20Hackathon.202021-07-30/near/247713553) while we were still in the prototyping phase, since it's difficult to iterate on top of. If that's true, documenting its ideas will remain important so they don't get lost to obscurity.

**Comment by ecstatic-morse** [comment]
## Intuition for `datafrog_opt`

This datalog program computes the same results (`errors` and `subset_errors`) as the "naive" version of the Polonius, but with less intermediate tuples.

### Redundant `errors`

Let's look at the naive version of `errors` (with the naive version of `loan_live_at` inlined into the rule body):

```prolog
errors(loan, node) :-
      loan_invalidated_at(loan, node),
      origin_contains_loan_on_entry(origin, loan, node),
      origin_live_on_entry(origin, node).
```

In plain english, a borrow error occurs when a loan is invalidated at a given point while contained by region that is live at that point. Notably, if the same loan is contained in multiple live regions while it is invalidated, we
don't gain any additional information: the same error will be generated multiple times.

For example, perhaps we have a two (non-placeholder) origins `'a` and `'b` such that `'a` is a subset of `'b` (`'a <: 'b`). If both `'a` and `'b` are live at some point in the control-flow graph, any borrow errors caused by a loan in `'a` will also be generated by that same loan in `'b`. For this reason, while both regions are live, there is no need to propagate loans from `'a` to `'b`. As long as each live loan is contained by at least one live region, we will emit a borrow error if that loan is invalidated.

In the past, this meant we could stop computing the full transitive closure of the `subset` relation as soon as we get to a live origin. Any edges from the live origin to its supersets can only create redundant borrow errors. The addition of placeholder region errors complicates things, since we need to derive all possible subset constraints between two placeholder origins, which
may pass through any number of non-placeholders. We deal with this by computing the `placeholder_subset` (all subset relationships that may involve a placeholder region) separately from the `subset` used to compute borrow errors.

Remember, all this is only valid while the regions on both sides of our subset constraint `'a <: 'b`. If `'a` (the subset) goes dead, we must propagate all of its loans to `'b` (the superset), guaranteeing that each live loan is contained in at least one live origin. If `'b` goes dead, we must propagate all loans in `'a` through `'b` to all of its live supersets (origins `'c` such that `'c` is
live and `'b <: 'c`). That way, if `'a` goes dead at a later point, its loans will be remembered.

**Comment by lqd** [comment]
👍 this also matches my intuition and the little documentation we have about it.

> We deal with this by computing the placeholder_subset (all subset relationships that may involve a placeholder region) separately from the subset used to compute borrow errors.

Since we only want to know the origins into which placeholders flow (to ultimately know all the placeholders that a placeholder can reach), "all subset relationships that may involve a placeholder region" may give the impression that non-placeholders flowing into a placeholder (since a placeholder is involved) are of interest here, but it's really that the subset is downstream from a placeholder that is the most important property. No big deal of course, especially for an aside like this.
